Data-Structures-Trie.md

```diff
@@ -0,0 +1,99 @@
+But when you use a trie for fetching data, the time complexity to get data is `O(w)`, where `w` is the length of the string.
+
+## How to implement a trie?
+
+A trie typically, looks something like this,
+
+![Trie](https://community.topcoder.com/i/education/alg_tries.png)
+
+The above image stores, assoc, algo, all, also, tree, trie.
+
+Let's implement a trie in python, for storing words with their meanings from english dictionary.
+
+```python
+ALPHABET_SIZE = 26 # For English
+
+class TrieNode:
+	def __init__(self):
+		self.edges = [None]*(ALPHABET_SIZE) # Each index respective to each character.
+		self.meaning = None # Meaning of the word.
+		self.ends_here = False # Tells us if the word ends here.
+```
+As you can see, edges are 26 in length, each index referring to each character in the alphabet. 'A' corresponding to 0, 'B' to 1, 'C' to 2 ... 'Z' to 25th index. If the character you are looking for is pointing to `None`, that implies the word is not there in the trie.
+
+A typical Trie should implement at least these two functions:
+	- add_word(word,meaning)
+	- search_word(word)
+Additionally, one can also add something like
+	- get_all_words()
+	- get_all_words_with_prefix(prefix)
+
+#### Adding Word to the trie
+
+```python
+	def add_word(self,word,meaning):
+		if len(word)==0:
+			self.ends_here = True # Because we have reached the end of the word
+			self.meaning = meaning # Adding the meaning to that node
+			return 
+		ch = word[0] # First character
+		# ASCII value of the first character (minus) the ASCII value of 'a'-> the first character of our ALPHABET gives us the index of the edge we have to look up.
+		index = ord(ch) - ord('a')
+		if self.edges[index] == None:
+			# This implies that there's no prefix with this character yet.
+			new_node = TrieNode()
+			self.edges[index] = new_node
+
+		self.edges[index].add(word[1:],meaning) #Adding the remaining word
+
+```
+
+#### Retrieving data
+
+```python
+	def search_word(self,word):
+		if len(word)==0 and self.ends_here:
+			return True
+		ch = word[0]
+		index = ord(ch)-ord('a')
+		if self.edge[index]== None:
+			return False
+		else:
+			return self.edge[index].search_word(word[1:])
+
+```
+
+The `search_word` function will tell us if the word exists in the Trie or not. Since ours is a dictionary, we need to fetch the meaning as well, now lets declare a function to do that.
+
+```python
+	def get_meaning(self,word):
+		if len(word)==0 and self.ends_here:
+			return self.meaning
+		ch = word[0]
+		index = ord(ch) - ord('a')
+		if self.edges[index] == None:
+			return "Word doesn't exist in the Trie"
+		else:
+			return self.edges[index].get_meaning(word[1:])
+```
+
+:rocket: [Run Code](https://repl.it/CWaJ)
```

Comment:
No, I mean, you use them in a main function for inserting, deleting, searching etc.